File: sdks/ts/memory/src/http/problem.ts
// SPDX-License-Identifier: Apache-2.0

/**
 * Problem+JSON helpers for the memory daemon. Mirrors the Go
 * `go/internal/httpd.Problem` shape so cross-SDK conformance assertions
 * see the same body on every error.
 *
 * The wire contract is documented in `spec/PROTOCOL.md` under "Error
 * shape" and "Error codes". A `Problem` body always carries `status`,
 * `title`, and usually a `code` plus `detail`.
 */

import {
  ErrConflict,
  ErrInvalidPath,
  ErrNotFound,
  ErrReadOnly,
  StoreError,
} from '../store/errors.js'

export type Problem = {
  readonly status: number
  readonly title: string
  readonly code?: string
  readonly detail?: string
}

export const CONTENT_TYPE_PROBLEM = 'application/problem+json'

export const problemResponse = (problem: Problem): Response => {
  const body: Record<string, unknown> = {
    status: problem.status,
    title: problem.title,
  }
  if (problem.code !== undefined) body.code = problem.code
  if (problem.detail !== undefined) body.detail = problem.detail
  return new Response(JSON.stringify(body), {
    status: problem.status,
    headers: { 'content-type': CONTENT_TYPE_PROBLEM },
  })
}

export const notFound = (detail: string): Response =>
  problemResponse({ status: 404, title: 'Not Found', code: 'not_found', detail })

export const validationError = (detail: string): Response =>
  problemResponse({
    status: 400,
    title: 'Bad Request',
    code: 'validation_error',
    detail,
  })

export const payloadTooLarge = (detail: string): Response =>
  problemResponse({
    status: 413,
    title: 'Payload Too Large',
    code: 'payload_too_large',
    detail,
  })

export const unauthorized = (detail: string): Response =>
  problemResponse({
    status: 401,
    title: 'Unauthorized',
    code: 'unauthorized',
    detail,
  })

export const forbidden = (detail: string): Response =>
  problemResponse({
    status: 403,
    title: 'Forbidden',
    code: 'forbidden',
    detail,
  })

export const conflict = (detail: string): Response =>
  problemResponse({
    status: 409,
    title: 'Conflict',
    code: 'conflict',
    detail,
  })

export const confirmationRequired = (detail: string): Response =>
  problemResponse({
    status: 428,
    title: 'Confirmation Required',
    code: 'confirmation_required',
    detail,
  })

export const internalError = (detail: string): Response =>
  problemResponse({
    status: 500,
    title: 'Internal Server Error',
    code: 'internal_error',
    detail,
  })

/**
 * Map a store-level error onto the Problem+JSON shape documented in
 * `spec/PROTOCOL.md`. Returns `undefined` when the error is unknown so
 * callers can fall back to `internalError`.
 */
export const storeProblem = (err: unknown): Response | undefined => {
  if (err instanceof ErrNotFound) {
    return notFound(err.message)
  }
  if (err instanceof ErrInvalidPath) {
    return validationError(err.message)
  }
  if (err instanceof ErrConflict) {
    return conflict(err.message)
  }
  if (err instanceof ErrReadOnly) {
    return conflict(err.message)
  }
  if (err instanceof StoreError) {
    return internalError(err.message)
  }
  return undefined
}

/**
 * Convenience JSON helper matching the Go daemon's writeJSON.
 */
export const jsonResponse = (status: number, body: unknown): Response =>
  new Response(JSON.stringify(body), {
    status,
    headers: { 'content-type': 'application/json' },
  })

File: go/internal/httpd/problem.go
// SPDX-License-Identifier: Apache-2.0

// Package httpd contains internal HTTP daemon helpers shared by cmd/memory
// subcommands. It defines the Problem+JSON encoder used across all
// handlers and the route registrar pattern lifted from the jeff server.
package httpd

import (
	"encoding/json"
	"errors"
	"net/http"

	"github.com/jeffs-brain/memory/go/brain"
)

// Problem is an RFC 7807-adjacent error body. Fields match the shape
// documented in spec/PROTOCOL.md.
type Problem struct {
	Status int    `json:"status"`
	Title  string `json:"title,omitempty"`
	Detail string `json:"detail,omitempty"`
	Code   string `json:"code,omitempty"`
}

// WriteProblem encodes p as application/problem+json and writes it.
func WriteProblem(w http.ResponseWriter, p Problem) {
	w.Header().Set("Content-Type", "application/problem+json")
	w.WriteHeader(p.Status)
	_ = json.NewEncoder(w).Encode(p)
}

// NotImplemented writes a 501 Problem+JSON body indicating the handler is
// not yet wired. Used by stub handlers that have not been ported.
func NotImplemented(w http.ResponseWriter) {
	WriteProblem(w, Problem{
		Status: http.StatusNotImplemented,
		Title:  "Not Implemented",
		Detail: "handler not wired",
		Code:   "not_implemented",
	})
}

// NotFound writes a 404 Problem+JSON.
func NotFound(w http.ResponseWriter, detail string) {
	WriteProblem(w, Problem{
		Status: http.StatusNotFound,
		Title:  "Not Found",
		Detail: detail,
		Code:   "not_found",
	})
}

// Conflict writes a 409 Problem+JSON.
func Conflict(w http.ResponseWriter, detail string) {
	WriteProblem(w, Problem{
		Status: http.StatusConflict,
		Title:  "Conflict",
		Detail: detail,
		Code:   "conflict",
	})
}

// ValidationError writes a 400 Problem+JSON with the validation_error
// code; used for body parsing failures and bad path shapes.
func ValidationError(w http.ResponseWriter, detail string) {
	WriteProblem(w, Problem{
		Status: http.StatusBadRequest,
		Title:  "Bad Request",
		Detail: detail,
		Code:   "validation_error",
	})
}

// PayloadTooLarge writes a 413 Problem+JSON.
func PayloadTooLarge(w http.ResponseWriter, detail string) {
	WriteProblem(w, Problem{
		Status: http.StatusRequestEntityTooLarge,
		Title:  "Payload Too Large",
		Detail: detail,
		Code:   "payload_too_large",
	})
}

// UnsupportedMediaType writes a 415 Problem+JSON.
func UnsupportedMediaType(w http.ResponseWriter, detail string) {
	WriteProblem(w, Problem{
		Status: http.StatusUnsupportedMediaType,
		Title:  "Unsupported Media Type",
		Detail: detail,
		Code:   "unsupported_media_type",
	})
}

// Unauthorized writes a 401 Problem+JSON.
func Unauthorized(w http.ResponseWriter, detail string) {
	WriteProblem(w, Problem{
		Status: http.StatusUnauthorized,
		Title:  "Unauthorized",
		Detail: detail,
		Code:   "unauthorized",
	})
}

// Forbidden writes a 403 Problem+JSON.
func Forbidden(w http.ResponseWriter, detail string) {
	WriteProblem(w, Problem{
		Status: http.StatusForbidden,
		Title:  "Forbidden",
		Detail: detail,
		Code:   "forbidden",
	})
}

// InternalError writes a 500 Problem+JSON with a generic title.
func InternalError(w http.ResponseWriter, detail string) {
	WriteProblem(w, Problem{
		Status: http.StatusInternalServerError,
		Title:  "Internal Server Error",
		Detail: detail,
		Code:   "internal_error",
	})
}

// WriteStoreError maps a brain.Store error to a Problem+JSON response
// using the canonical sentinel-to-code mapping. Returns true when err
// matched a known sentinel and the response has been written; callers
// should fall back to [InternalError] otherwise.
func WriteStoreError(w http.ResponseWriter, err error) bool {
	if err == nil {
		return false
	}
	switch {
	case errors.Is(err, brain.ErrNotFound):
		NotFound(w, err.Error())
	case errors.Is(err, brain.ErrConflict):
		Conflict(w, err.Error())
	case errors.Is(err, brain.ErrInvalidPath), errors.Is(err, brain.ErrValidation):
		ValidationError(w, err.Error())
	case errors.Is(err, brain.ErrPayloadTooLarge):
		PayloadTooLarge(w, err.Error())
	case errors.Is(err, brain.ErrUnsupportedMedia):
		UnsupportedMediaType(w, err.Error())
	case errors.Is(err, brain.ErrUnauthorized):
		Unauthorized(w, err.Error())
	case errors.Is(err, brain.ErrForbidden):
		Forbidden(w, err.Error())
	case errors.Is(err, brain.ErrReadOnly):
		WriteProblem(w, Problem{Status: http.StatusServiceUnavailable, Title: "Read Only", Detail: err.Error(), Code: "read_only"})
	default:
		return false
	}
	return true
}
